lib/puppet/util/tag_set.rb

```diff
@@ -16,6 +16,7 @@ def self.from_data_hash(data)
     self.new(data)
   end
 
+  # TODO: A method named #to_data_hash should not return an array
   def to_data_hash
     to_a
```

Comment:
Generally speaking, this does't work. If someone tries to serialize/deserialize a `TagSet` as a stand-alone entity, it will fail. The only reason it does cause problems now is that the array is embedded in serialization of a `Resource` and that class expects an array of tags when deserializing.

The resolution is to let the `Resource` continue to serialize the tags in an `Array` (by calling `TagSet#to_a` directly), and then let this method return a Hash that describes both its type and its set of tags (to make it possible to serialize tagsets in places where the owner doesn't fix the problem). I think that's beyond the scope of this ticket and would like to do that as part of implementing my proposed serialization.